### PR TITLE
Improved Google OIDC connector

### LINKED ISSUE #5521: Provide ability to pull multiple domain groups from Google Workspace (formerly gsuite)
### Feature Request

Provide the ability to pull all or specific domain user groups within a Google Workspace.  Currently only the user's specific @domain groups will be retrieved.  
Ex:
`steven@me.com` in groups: `developer@me.com, admin@me.com, deploy@app.me.com, support@support.me.com`

Only `developer@me.com`, and `admin@me.com` will be retrieved as claims.

If the user was `steven@app.me.com` then only `deploy@app.me.com` will be retrieved as a claim.

### Motivation
Users may be assigned to multiple groups that use different sub-domains then the user.  This allows for using all the group assignments for role to claim mapping.

### Who's it for?
Pro, Enterprise, Cloud

### LINKED ISSUE #8122: Google SSO nested groups
### Description

Google nested group are not expanded when fetching groups using our oauth code  
TLDR: when user is a member of a group that is a subgroup of another gmail, it’s not reflected in the groups arriving in SSO.

### PR DESCRIPTION
**Pull request by espadolini**
Optional Google Cloud Identity support to fetch transitive groups (i.e. groups that a user belongs to both directly and indirectly through nested groups) for SSO.

This bumps the version of the `oidc` resources from `v2` to `v3`, supporting both versions, with no change to the actual spec for the resource; the behavior of `v2` connectors will stay the same, while `v3` connectors will not filter the returned groups by domain name, and (depending on the permissions of the configured service account) will attempt to fetch both direct and indirect groups.

Fixes #8122.
Fixes #5521.

### CODE REVIEW

**Review by rosstimothy — COMMENTED** [review]
Can you add test coverage? 

Either of the following would allow you to test with out needing to talk to google:
 - Migrating the `groupsFromGsuite*` functions to an interface would allow you to mock the client
 - Utilizing [WithEndpoint](https://pkg.go.dev/google.golang.org/api@v0.65.0/option#WithEndpoint) or [WithGRPCConn](https://pkg.go.dev/google.golang.org/api@v0.65.0/option#WithGRPCConn) would allow you to mock google

**Review by ibeckermayer — APPROVED** [review]

File: lib/auth/oidc_test.go [hunk 1]
```diff
@@ -176,3 +184,114 @@ func (s *fakeIDP) configurationHandler(w http.ResponseWriter, r *http.Request) {
...
+		email      string
+		transitive bool
+		groups     []string
+	}{
+		{"alice@foo.example", true, []string{"group1@foo.example", "group2@sub.foo.example", "group3@bar.example", "group4@bar.example"}},
+		{"alice@foo.example", false, []string{"group1@foo.example", "group2@sub.foo.example", "group3@bar.example"}},
+		{"bob@foo.example", false, []string{"group1@foo.example"}},
+		{"bob@foo.example", true, []string{"group1@foo.example"}},
+	} {
+		connector.(*types.OIDCConnectorV2).Spec.GoogleTransitiveGroups = testCase.transitive
```
Comment (rosstimothy) [inline lib/auth/oidc_test.go:289]: You could create a connector per test case and set this during instantiation to remove the type assertion.

File: constants.go [hunk 2]
```diff
@@ -547,17 +547,6 @@ const (
 	TraitInternalDBUsersVariable = "{{internal.db_users}}"
 )
 
-const (
-	// GSuiteIssuerURL is issuer URL used for GSuite provider
-	GSuiteIssuerURL = "https://accounts.google.com"
```
Comment (rosstimothy) [inline constants.go:552]: Seems like you could keep this one?

File: lib/auth/oidc_google.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+			}
+			return nil
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return groups, nil
+}
+
+func groupsFromGsuiteCloudidentity(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
```
Comment (rosstimothy) [inline lib/auth/oidc_google.go:164]:
```suggestion
func groupsFromGsuiteCloudIdentity(ctx context.Context, email string, clientOptions ...option.ClientOption) ([]string, error) {
```

File: lib/auth/oidc_test.go [hunk 4]
```diff
@@ -176,3 +184,114 @@ func (s *fakeIDP) configurationHandler(w http.ResponseWriter, r *http.Request) {
...
+
+		resp := &directory.Groups{}
+		for _, groupEmail := range directGroups[email] {
+			resp.Groups = append(resp.Groups, &directory.Group{Email: groupEmail})
+		}
+
+		b, err := json.Marshal(resp)
+		require.NoError(t, err)
+
+		rw.Write(b)
```
Comment (rosstimothy) [inline lib/auth/oidc_test.go:224]:
```suggestion
		require.NoError(t, json.NewEncoder(rw).Encode(resp))
```

File: lib/auth/oidc_test.go [hunk 5]
```diff
@@ -176,3 +184,114 @@ func (s *fakeIDP) configurationHandler(w http.ResponseWriter, r *http.Request) {
...
+					},
+					RelationType: relationType,
+				})
+			}
+		}
+
+		b, err := json.Marshal(resp)
+		require.NoError(t, err)
+
+		rw.Write(b)
```
Comment (rosstimothy) [inline lib/auth/oidc_test.go:262]:
```suggestion
		require.NoError(t, json.NewEncoder(rw).Encode(resp))
```

File: lib/auth/oidc_google.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+		return nil, trace.Wrap(err)
+	}
+
+	var groups []string
+	err = service.Groups.Memberships.SearchTransitiveGroups("groups/-").
+		// the google API docs claim that the query string is a CEL expression
+		// (https://opensource.google/projects/cel) but the call will fail if
+		// you use double quotes instead of single quotes in spite of them being
+		// equivalent according to the CEL specs
+		Query(fmt.Sprintf("member_key_id == '%s' && 'cloudidentity.googleapis.com/groups.discussion_forum' in labels", email)).
```
Comment (rosstimothy) [inline lib/auth/oidc_google.go:176]: Where did you get the values for the `SearchTransitiveGroups` and `Query`? Could you add a comment for those of us that aren't as familiar with the cloud identity api as you have become?

**Review by rosstimothy — APPROVED** [review]
Let a few small comments. Looks good otherwise.

File: constants.go (hunk 2 above)
Comment (espadolini) [reply constants.go:552]: Just there in `constants.go`, or private in `lib/auth`?

File: lib/auth/oidc.go [hunk 7]
```diff
@@ -828,76 +685,9 @@ func (a *Server) getClaims(oidcClient *oidc.Client, connector types.OIDCConnecto
...
-		} else {
-			if gsuiteClaims != nil {
-				log.Debugf("Got gsuiteClaims claims from Google Workspace: %v.", gsuiteClaims)
-			}
-			claims, err = mergeClaims(claims, gsuiteClaims)
-			if err != nil {
-				return nil, trace.Wrap(err)
-			}
-		}
+	claims, err = addGsuiteClaims(a.closeCtx, connector, claims)
```
Comment (russjones) [inline lib/auth/oidc.go:688]: I think it's called Google Workspace now, so how about `addGoogleWorkspaceClaims`.

File: lib/auth/oidc_google.go [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+// connector. The credentials in the default case are loaded from the connector
+// with the getGsuiteCredentialsForConnector function.
+func addGsuiteClaims(ctx context.Context, connector types.OIDCConnector, claims jose.Claims, clientOptions ...option.ClientOption) (jose.Claims, error) {
+	// If google_service_account_uri and google_service_account are not set, we
+	// assume that this is a non-GWorkspace OIDC provider using the same
+	// issuer URL as Google Workspace (e.g.
+	// https://developers.google.com/identity/protocols/oauth2/openid-connect).
+	if connector.GetIssuerURL() != gsuiteIssuerURL || (connector.GetGoogleServiceAccountURI() == "" && connector.GetGoogleServiceAccount() == "") {
+		return claims, nil
+	}
```
Comment (russjones) [inline lib/auth/oidc_google.go:59]: I'd wrap this into it's own function `isGoogleWorkspace` then you can update `getClaims` to have something like the following:

```go
// If the OIDC provider is configured to be Google Workspace, add in addition claims.
if ok := isGoogleWorkspace(...); ok {
    claims, err = addGoogleWorkspaceClaims(...)
    if err != nil {
        return nil, trace.Wrap(err)
    }
}
return claims, nil
```

File: lib/auth/oidc_google.go [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+	// message.
+	//
+	// The query string was lifted directly from
+	// https://cloud.google.com/identity/docs/how-to/query-memberships#searching_for_all_group_memberships_of_a_member
+	// and some more informations on group labels can be found at
+	// https://cloud.google.com/identity/docs/groups#group_properties . The
+	// actual docs for the API call are at
+	// https://cloud.google.com/identity/docs/reference/rest/v1/groups.memberships/searchTransitiveGroups .
+	err = service.Groups.Memberships.SearchTransitiveGroups("groups/-").
+		Query(fmt.Sprintf("member_key_id == '%s' && 'cloudidentity.googleapis.com/groups.discussion_forum' in labels", email)).
```
Comment (russjones) [inline lib/auth/oidc_google.go:198]: How does this query language work? Does it have the equivalent of parameterized queries?

I'm wondering if you can do something like SQL injection with it?

**Comment by russjones** [comment]
@espadolini Just to make sure I understand this correctly, this PR solves two problems.

1. Fetch all domain groups for a user.
2. Fetch transitive group membership for a user.

For (1), the issue is resolved by no longer filtering on a particular domain when we make the [groups.list](https://developers.google.com/admin-sdk/directory/reference/rest/v1/groups/list) query.

For (2), from what I can tell the main issue is that to support getting transitive group membership, we have to call different API endpoints that not all users may have access to due to needing to be on a [Google Workspace Enterprise plan].

Is there any other way to solve (2) or is the Enterprise plan the only way to get transitive group membership for a user?

File: lib/auth/oidc_google.go (hunk 9 above)
Comment (espadolini) [reply lib/auth/oidc_google.go:198]: I tried looking for a "proper" way to do this, but the only library to handle this (https://github.com/google/cel-go) is apparently only for evaluating CEL expressions in text form, not for building them; I don't think it's a problem here, because the `email` parameter we use to format the string is the `email` that came as part of the OIDC claims, so it's something that Google considers to be a valid email address - which I don't believe can include any `'`s.

**Comment by espadolini** [comment]
That's correct. The change in default behavior could be mitigated by putting the groups that don't match the domain of the user into a separate claim (`foreign_groups`?) - to maintain the old behavior with the new API we have to do the filtering ourselves, anyway.

I don't think there's a reasonable way to get transitive group memberships without using the intended API - anything ad-hoc that we build ourselves would be too spammy and fragile.

**Comment by russjones** [comment]
@espadolini For (1), let's bump the version of the connector to `OIDCConnectorV3` and in this version drop the domain filter. This way anyone using the old connector will continue to get domain filtering and no unexpected behavior changes but if you upgrade you get all groups.

For (2), can we update the logic to always call `service.Groups.Memberships.SearchTransitiveGroups` first and if it fails, then fallback to `service.Groups.List`? This way we don't have to add an additional field and will still get the same functionality.

**Comment by espadolini** [comment]
> @espadolini For (1), let's bump the version of the connector to `OIDCConnectorV3` and in this version drop the domain filter. This way anyone using the old connector will continue to get domain filtering and no unexpected behavior changes but if you upgrade you get all groups.

Ack.

> For (2), can we update the logic to always call `service.Groups.Memberships.SearchTransitiveGroups` first and if it fails, then fallback to `service.Groups.List`? This way we don't have to add an additional field and will still get the same functionality.

Even before calling the actual API we can check if the specified service account is allowed to use the `CloudIdentityGroupsReadonlyScope` scope; if that succeeds we can assume that the system should use the new API, otherwise it should use the old one (and if we also do the check for the `AdminDirectoryGroupReadonlyScope` scope we get a better error message).

I'd be wary of just blindly falling back to the old API on any spurious failure of the new, because if there's roles associated with transitive groups it could lead to a very inconsistent experience for the users - then again, we could just document that for consistent behavior, the service account should only have one of the two scopes enabled. What do you think @russjones?

**Comment by russjones** [comment]
@espadolini Please add documentation in another PR on how scopes work now (transitive or regular).

**Review by rosstimothy — APPROVED** [review]

File: lib/auth/oidc_google.go [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+		}
+
+		if credentials != nil {
+			log.Debugf("fetching transitive Google groups for %v", email)
+			googleGroups, err = groupsFromGoogleCloudIdentity(ctx, email, option.WithCredentials(credentials))
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+		} else {
+			credentials, err := getGoogleWorkspaceCredentials(ctx, connector, directory.AdminDirectoryGroupReadonlyScope)
```
Comment (ibeckermayer) [inline lib/auth/oidc_google.go:83]: Similar logic to the V2 logic below, consider parameterizing into a function.

File: lib/services/oidc.go [hunk 11]
```diff
@@ -102,8 +102,8 @@ func UnmarshalOIDCConnector(bytes []byte, opts ...MarshalOption) (types.OIDCConn
 		return nil, trace.Wrap(err)
 	}
 	switch h.Version {
-	case types.V2:
-		var c types.OIDCConnectorV2
+	case types.V2, types.V3:
```
Comment (ibeckermayer) [inline lib/services/oidc.go:105]: Perhaps this is standard practice elsewhere in the codebase, in which case ignore this comment, however if not then it may be confusing why we are using an `OIDCConnectorV3` for `case V2` and worthy of a clarifying comment.

**Review by ibeckermayer — APPROVED** [review]

File: lib/services/oidc.go (hunk 11 above)
Comment (espadolini) [reply lib/services/oidc.go:105]: I have a comment in `(*OIDCConnectorV3).CheckAndSetDefaults()` already, but it's probably useful to restate it here as well.

File: lib/auth/oidc_google.go (hunk 10 above)
Comment (espadolini) [reply lib/auth/oidc_google.go:83]: I had tried refactoring it a bit but this is still the clearest way to write it that I've found, I'm afraid. 😢
